pkg/controller/elasticsearch/driver/upscale_state.go

```diff
@@ -10,58 +10,79 @@ import (
 	appsv1 "k8s.io/api/apps/v1"
 	corev1 "k8s.io/api/core/v1"
 
-	"github.com/elastic/cloud-on-k8s/pkg/apis/elasticsearch/v1beta1"
 	"github.com/elastic/cloud-on-k8s/pkg/controller/common"
 	"github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/label"
 	"github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/nodespec"
 	"github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/sset"
 	"github.com/elastic/cloud-on-k8s/pkg/utils/k8s"
 )
 
-type upscaleStateBuilder struct {
-	once         *sync.Once
-	upscaleState *upscaleState
-}
-
-func (o *upscaleStateBuilder) InitOnce(c k8s.Client, es v1beta1.Elasticsearch, esState ESState) (*upscaleState, error) {
-	if o.once == nil {
-		o.once = &sync.Once{}
-	}
-	var err error
-	o.once.Do(func() {
-		o.upscaleState, err = newUpscaleState(c, es, esState)
-	})
-	return o.upscaleState, err
-}
-
 type upscaleState struct {
 	isBootstrapped      bool
 	allowMasterCreation bool
+	recordedCreates     int32
+	createsAllowed      *int32
+	ctx                 upscaleCtx
+	once                *sync.Once
 }
 
-func newUpscaleState(c k8s.Client, es v1beta1.Elasticsearch, esState ESState) (*upscaleState, error) {
-	state := &upscaleState{
-		isBootstrapped:      AnnotatedForBootstrap(es),
-		allowMasterCreation: true,
+func newUpscaleState(
+	ctx upscaleCtx,
+	actualStatefulSets sset.StatefulSetList,
+	expectedResources nodespec.ResourcesList,
+) *upscaleState {
+	return &upscaleState{
+		once: &sync.Once{},
+		ctx:  ctx,
+		createsAllowed: calculateCreatesAllowed(
+			ctx.es.Spec.UpdateStrategy.ChangeBudget.GetMaxSurgeOrDefault(),
+			actualStatefulSets.ExpectedNodeCount(),
+			expectedResources.StatefulSets().ExpectedNodeCount()),
 	}
-	if !state.isBootstrapped {
-		return state, nil
+}
+
+func (s *upscaleState) buildOnce() (result error) {
+	if s.once != nil {
+		s.once.Do(func() {
+			s.isBootstrapped = AnnotatedForBootstrap(s.ctx.es)
+			s.allowMasterCreation = true
+
+			if s.isBootstrapped {
+				// is there a master node creation in progress already?
+				masters, err := sset.GetActualMastersForCluster(s.ctx.k8sClient, s.ctx.es)
+				if err != nil {
+					result = err
+					return
```

Comment:
It feels a bit easier to read to me if we don't use the named return value.
`return err` directly and `return nil` at the bottom?

Reply:
Can't as this func is an argument to `once.Do` which doesn't expect a func that returns a value.

Reply:
But linter insisted on not using named return value, so changed it anyway.